src/cloudWatchLogs/registry/logDataRegistry.ts

```diff
@@ -120,26 +121,60 @@ export class LogDataRegistry {
             getLogger().debug(`No registry entry for ${uri.path}`)
             return
         }
-        const nextToken = headOrTail === 'head' ? logData.previous?.token : logData.next?.token
+        const request: CloudWatchLogsResponse = {
+            events: [],
+            nextForwardToken: logData.next?.token,
+            nextBackwardToken: logData.previous?.token,
+        }
+
+        const isHead = headOrTail === 'head'
+        const stream = pageableToCollection(
+            (r: typeof request) =>
+                logData.retrieveLogsFunction(
+                    logData.logGroupInfo,
+                    logData.parameters,
+                    isHead ? r.nextBackwardToken : r.nextForwardToken
+                ),
+            request,
+            isHead ? 'nextBackwardToken' : 'nextForwardToken'
+        )
 
-        // TODO: Consider getPaginatedAwsCallIter? Would need a way to differentiate between head/tail...
-        const logEvents = await logData.retrieveLogsFunction(logData.logGroupInfo, logData.parameters, nextToken)
+        async function firstOrLast<T>(
+            iterable: AsyncIterable<T>,
+            predicate: (item: T) => boolean
+        ): Promise<T | undefined> {
+            let last: T | undefined
+            for await (const item of iterable) {
+                if (predicate((last = item))) {
+                    return item
+                }
+            }
+            return last
+        }
+
+        const responseData = await firstOrLast(stream, resp => resp.events.length > 0)
+
+        if (!responseData) {
+            return
+        }
 
         const newData =
             headOrTail === 'head'
-                ? (logEvents.events ?? []).concat(logData.data)
-                : logData.data.concat(logEvents.events ?? [])
+                ? (responseData.events ?? []).concat(logData.data)
+                : logData.data.concat(responseData.events ?? [])
 
         const tokens: Pick<CloudWatchLogsData, 'next' | 'previous'> = {}
         // update if no token exists or if the token is updated in the correct direction.
         if (!logData.previous || headOrTail === 'head') {
-            tokens.previous = {
-                token: logEvents.nextBackwardToken ?? '',
+            const token = responseData.nextBackwardToken
```

Comment:
I'm not that familiar with the cloudwatch logs code, but is there any reason we don't have: 
```suggestion
            const token = responseData.nextBackwardToken ?? request.nextBackwardToken
```
similar to what we have on line 175?

Reply:
When we go "backward" we are going back in time (each page we visit is guaranteed to exist), so if a response has no backward token it implies we are at the beginning and are done.

This is different from going "forward" since a new page can eventually exist, which is why we store `request.nextForwardToken` if there is no new token from the response. When a response has no `nextForwardToken` it implies it is the tail and at some point it can have new data.